File: Haskell/eu003.hs
sqrtInt :: Int -> Int
sqrtInt = floor . sqrt . fromIntegral

primes :: [Int]
primes = 2 : [x | x <- [3, 5..]
                , and [x `rem` y /= 0 | y <- takeWhile (<= sqrtInt x) primes]]

factorize :: Int -> [Int] -> Int
factorize n (p:ps)
  | n == p = p
  | n `rem` p == 0 = factorize (n `quot` p) ps
  | otherwise = factorize n ps

main :: IO ()
main = print (factorize 600851475143 primes)

File: Clojure/eu003.clj
(defn primes
  ([] (cons 2 (lazy-seq (primes 3 [2]))))
  ([x, ps]
    (if (every? #(not= 0 (rem x %)) ps)
        (cons x (lazy-seq (primes (+ x 2) (conj ps x))))
        (primes (+ x 2) ps))))

(defn factorize [n, ps]
  (let [[p & ps'] ps]
    (if (= n p)
        n
        (if (= 0 (rem n p))
            (factorize (quot n p) ps')
            (factorize n ps')))))

(println (factorize 600851475143 (primes)))
